pkg/syncer/namespace/resources.go

```diff
@@ -0,0 +1,151 @@
+
+	lock sync.Mutex // use lock instead of deep-copying objects.
+)
+
+func init() {
+	loadObjectOrDie("serviceaccount_dns.yaml", &serviceAccount)
+	loadObjectOrDie("role_dns.yaml", &role)
+	loadObjectOrDie("rolebinding_dns.yaml", &roleBinding)
+	loadObjectOrDie("deployment_dns.yaml", &deployment)
+	loadObjectOrDie("service_dns.yaml", &service)
+}
+
+func createServiceAccountFn(kubeclient kubernetes.Interface, namespace string) func(context.Context, string) error {
+	return func(ctx context.Context, name string) error {
+		lock.Lock()
+		defer lock.Unlock()
+
+		serviceAccount.Name = name
+		serviceAccount.Namespace = namespace
+
+		_, err := kubeclient.CoreV1().ServiceAccounts(namespace).Create(ctx, &serviceAccount, metav1.CreateOptions{})
+		return err
+	}
+}
+
+func createRoleFn(kubeclient kubernetes.Interface, namespace string) func(context.Context, string) error {
+	return func(ctx context.Context, name string) error {
+		lock.Lock()
+		defer lock.Unlock()
+
+		role.Name = name
+		role.Namespace = namespace
+		role.Rules[0].ResourceNames[0] = name
+
+		_, err := kubeclient.RbacV1().Roles(namespace).Create(ctx, &role, metav1.CreateOptions{})
+		return err
+	}
+}
+
+func createRoleBindingFn(kubeclient kubernetes.Interface, namespace string) func(context.Context, string) error {
+	return func(ctx context.Context, name string) error {
+		lock.Lock()
+		defer lock.Unlock()
+
+		roleBinding.Name = name
+		roleBinding.Namespace = namespace
+		roleBinding.RoleRef.Name = name
+		roleBinding.Subjects[0].Name = name
+		roleBinding.Subjects[0].Namespace = namespace
+
+		_, err := kubeclient.RbacV1().RoleBindings(namespace).Create(ctx, &roleBinding, metav1.CreateOptions{})
+		return err
+	}
+}
+
+func createDeploymentFn(kubeclient kubernetes.Interface, namespace, image string) func(context.Context, string) error {
+	return func(ctx context.Context, name string) error {
+		lock.Lock()
+		defer lock.Unlock()
+
+		deployment.Name = name
+		deployment.Namespace = namespace
+		deployment.Spec.Selector.MatchLabels["app"] = name
+		deployment.Spec.Template.Labels["app"] = name
+		deployment.Spec.Template.Spec.Containers[0].Image = image
+		deployment.Spec.Template.Spec.Containers[0].Args[3] = name
+		deployment.Spec.Template.Spec.ServiceAccountName = name
+
+		_, err := kubeclient.AppsV1().Deployments(namespace).Create(ctx, &deployment, metav1.CreateOptions{})
+		return err
+	}
+}
+
+func createServiceFn(kubeclient kubernetes.Interface, namespace string) func(context.Context, string) error {
+	return func(ctx context.Context, name string) error {
+		lock.Lock()
+		defer lock.Unlock()
+
+		service.Name = name
+		service.Namespace = namespace
+		service.Labels["app"] = name
+		service.Spec.Selector["app"] = name
+
+		_, err := kubeclient.CoreV1().Services(namespace).Create(ctx, &service, metav1.CreateOptions{})
+		return err
+	}
+}
+
+// load a YAML resource into a typed kubernetes object.
+func loadObjectOrDie(filename string, obj interface{}) {
```

Comment:
Why not have templates loaded from the disk at start here, and feed them with parameters + decode them as object during the creation ?

Reply:
the main reason is performance. Template instantiation is expensive. I found this approach a good compromise between readability (YAML is better than Go for representing k8s resources) and type safety.